draft-ietf-ppm-dap.md

```diff
@@ -1514,217 +1609,235 @@ to the next phase of the aggregation protocol.
-advance the aggregation job to the round the Leader has just reached by sending
-the new `prep_msg` message to the Helper in a POST request to the aggregation
-job URI used during initialization (see {{leader-init}}). The body of the
-request is an `AggregationJobContinueReq`:
+Next, the Leader sends a POST request to the aggregation job URI used during
+initialization (see {{leader-init}}) with media type
+"application/dap-aggregation-job-continue-req" and body structured as:
 
 ~~~
 struct {
-  u16 round;
+  u16 step;
   PrepareStep prepare_steps<1..2^32-1>;
 } AggregationJobContinueReq;
 ~~~
 
-The `round` field is the round of VDAF preparation that the Leader just reached
-and wants the Helper to advance to.
+The `step` field is the step of DAP aggregation that the Leader just reached
+and wants the Helper to advance to. The `prepare_steps` field is the sequence
+of prep steps constructed in the previous step. The `PrepareStep`s MUST be in
+the same order as the previous aggregate request.
 
-The `prepare_steps` field MUST be a sequence of `PrepareStep`s in the
-`continued` state containing the corresponding `inbound` prepare message.
+The Helper's response will be an `AggregationJobResp` message (see
+{{aggregation-helper-init}}), which the Leader validates according to the
+criteria in {{aggregation-job-validation}}. If the message is invalid, the
+Leader MUST abort the aggregation job.
 
-The media type is set to "application/dap-aggregation-job-continue-req".
+Otherwise, the Leader proceeds as follows with each report:
 
-The Helper's response will be an `AggregationJobResp` message (see
-{{aggregation-helper-init}}), which the Leader validates according to
-the criteria in {{aggregation-job-validation}}. If the message is valid, the
-Leader moves to the next round of continuation with the enclosed prepare steps.
-Otherwise, the Leader should abandon the aggregation job entirely.
+1. If the inbound prep step type is "continue" and the state is
+   `Continued(prep_state)`, then the Leader computes
+
+   ~~~
+   (state, outbound) = VDAF.ping_pong_req(agg_param, state, inbound)
+   ~~~
+
+   where `inbound` is the message payload. If `outbound != None`, then the
+   Leader stores `state` and `outbound` and proceeds to another continuation
+   step. If `outbound == None`, then the preparation process is complete: either
+   `state == Rejected()`, in which case the Leader rejects the report and
+   removes it from the candidate set; or `state == Finished(out_share)`, in
+   which case preparation is complete and the Leader stores the output share for
+   use in the collection sub-protocol {{collect-flow}}.
+
+1. Else if the type is "finished" and `state == Finished(out_share)`, then
+   preparation is complete and the Leader stores the output share for use in
+   the collection flow ({{collect-flow}}).
+
+1. Else if the type is "reject", then the Leader rejects the report and removes
+   it from the candidate set.
```

Comment:
Do we also need the exception here for `report_too_early` to match that in `{{leader-init}}`, or do we claim that this is impossible because any too early reports would have been rejected previously?

Reply:
I think it's the latter. (An implementation could guard against internal clock skew by adding the check here too, I suppose, but I think that's best left to implementations.)

Reply:
Good catch. In the absence of a MUST or MUST NOT stating otherwise, `report_too_early` is valid here. I would copy the language here.